### ISSUE TITLE
kata-shim memory leak

### ISSUE THREAD

**Issue by yiya1989** [kata-containers/kata-containers#5738]
The metrics function has collect all the kata-shim's net-ns netdev metrics, but the kata-shim net-ns is the hostnetwork.

`updateShimMetrics` and `UpdateRuntimeMetrics` both has this code as bellow, but `proc.NetDev()` equals to `/proc/self/net/dev` and due to the kata-shim belongs to the hostnetwork, then the `mutils.SetGaugeVecNetDev` will record all interface's metrics in the host network namespace.   The `kata_shim_netdev` only be set and update, but without reset, so when host network interface changed, then the `kata_shim_netdev` will continue increase which will result in memory continue inscrease. 
```
	proc, err := procfs.Self()
	if err != nil {
		return err
	}
        . . .
        // network device metrics
        if netdev, err := proc.NetDev(); err == nil {
	        // netdev: map[string]NetDevLine
	        for _, v := range netdev {
		        mutils.SetGaugeVecNetDev(katashimNetdev, v)
	        }
        }
```

Here is a metrics sample:
```
root@devbox-gaea:~# curl -fsS http://./metrics --unix-sock /run/vc/sbs/d29b349e5b24fb1da589942479aa6ce0e4b717f5c8020a252bd6d8c33a7e22bc/shim-monitor.sock |grep kata_shim_netdev |grep recv_compressed
kata_shim_netdev{interface="br-int",item="recv_compressed"} 0
kata_shim_netdev{interface="cni0",item="recv_compressed"} 0
kata_shim_netdev{interface="docker0",item="recv_compressed"} 0
kata_shim_netdev{interface="eth0",item="recv_compressed"} 0
kata_shim_netdev{interface="flannel.1",item="recv_compressed"} 0
kata_shim_netdev{interface="lo",item="recv_compressed"} 0
kata_shim_netdev{interface="lxcbr0",item="recv_compressed"} 0
kata_shim_netdev{interface="ovs-system",item="recv_compressed"} 0
kata_shim_netdev{interface="veth01e7bc92",item="recv_compressed"} 0
kata_shim_netdev{interface="veth10db999e",item="recv_compressed"} 0
kata_shim_netdev{interface="veth140df615",item="recv_compressed"} 0
kata_shim_netdev{interface="veth3db8b3e8",item="recv_compressed"} 0
kata_shim_netdev{interface="veth5154c3d0",item="recv_compressed"} 0
kata_shim_netdev{interface="veth5c5ef85a",item="recv_compressed"} 0
kata_shim_netdev{interface="veth66dfdef0",item="recv_compressed"} 0
kata_shim_netdev{interface="vethb68ae789",item="recv_compressed"} 0
```

**Comment by liubin** [comment]
Not sure `GaugeVec.Reset()` can fix this, but value to try it.

**Comment by yiya1989** [comment]
> Not sure `GaugeVec.Reset()` can fix this, but value to try it.

I agree with you that the GaugeVec.Reset() may fix this too, but is the kata-shim necessary to collect all net interface metrics in hostnetwork?
If it's more usefull to collect interface's metrics in the pod's network namespace?

**Comment by gkurz** [comment]
> `updateShimMetrics` and `UpdateRuntimeMetrics` both has this code as bellow, but `proc.NetDev()` equals to `/proc/self/net/dev`

This is true for `updateShimMetrics` but not for `UpdateRuntimeMetrics` which does :
```
	pids := s.hypervisor.GetPids()
	if len(pids) == 0 {
		return nil
	}

	hypervisorPid := pids[0]

	proc, err := procfs.NewProc(hypervisorPid)
	if err != nil {
		return err
	}
```

**Comment by littlejawa** [comment]
I've been experimenting on this topic, and found the following:

- using GaugeVec.Reset() definitely helps: we keep collecting more metrics than we should, but we don't make it grow forever by keeping track of removed/obsolete interfaces.
- I tried to collect the metrics from the network namespace, but it doesn't change anything.
  Maybe I'm not using it correctly: I have tried calling the metrics gathering function within a call to "EnterNetNS", but the reported interfaces are just the same.
  If anybody have any suggestion on that topic, it is welcome.

Now I have another question: do we really need to report the netdev metrics from the shim?
We already have `kata_hypervisor_netdev` repoted by the shim for the hypervisor process, and `kata_guest_netdev_stat` reported by the agent.
Both of them report the content of `/proc/.../net/dev`, limited to the interfaces they can see and use.

If I'm not mistaken, the shim doesn't use the network. Its communication to other processes is done throught unix sockets. Even for peer-pods, it talks to the cloud-api-adaptor which does the network part.
If the shim doesn't do anything with network, and the network-using processes are already reporting metrics of their usage, what information are we adding with the `kata_shim_netdev` metrics?
If we can identify what useful information the shim is the only one to see, maybe we can filter this out, and report only that?
And if the only thing we add is information about interfaces that we're not supposed to see and use, then maybe we should just stop reporting that metric altogether?
